engine/apps/alerts/incident_log_builder/incident_log_builder.py

```diff
@@ -10,26 +10,69 @@
 if typing.TYPE_CHECKING:
     from django.db.models.manager import RelatedManager
 
+    from apps.alerts.escalation_snapshot.snapshot_classes import EscalationPolicySnapshot, EscalationSnapshot
     from apps.alerts.models import AlertGroup, AlertGroupLogRecord, ResolutionNote
     from apps.base.models import UserNotificationPolicy, UserNotificationPolicyLogRecord
+    from apps.slack.models import SlackTeamIdentity
     from apps.user_management.models import User
 
 
+class PlanLines(typing.TypedDict):
+    plan_lines: typing.List[str]
+
+
+class NotificationPlanLines(PlanLines):
+    user_id: int
+    plan_lines: typing.List[str]
+    """
+    rendered notification policy line
+    """
+    is_the_first_notification_step: bool
+
+
+EscalationPlan = typing.Dict[timezone.timedelta, typing.List[PlanLines]]
+"""
+`dict` with `timedelta` as a key and values containing a `list` of escalation plan lines
+"""
+
+NotificationPlan = typing.Dict[timezone.timedelta, typing.List[NotificationPlanLines]]
+"""
+`dict` with `timedelta` as a key and values containing a `list` of notification plan lines
+"""
+
+MixedEscalationNotificationPlan = typing.Dict[
+    timezone.timedelta, typing.List[typing.Union[PlanLines, NotificationPlanLines]]
+]
+"""
+NOTE: this seems janky, maybe there's a better way for us to do this than jam two different object types into the same
+list? Opportunity for future refactoring...
+"""
+
+FlattendEscalationPlan = typing.Dict[timezone.timedelta, typing.List[str]]
+
+LogRecords = typing.List[typing.Union["AlertGroupLogRecord", "ResolutionNote", "UserNotificationPolicyLogRecord"]]
+
+UsersToNotify = typing.Iterable["User"]
+"""
+`typing.Iterable` ([docs](https://docs.python.org/3/library/collections.abc.html#collections.abc.Iterable)):
+
+ABC for classes that provide the __iter__() method
+
+We just need this because some methods for fetching users return `QuerySet`s, others simply assign a list. In the end
+we simply iterate over the users.
+"""
+
+
 class IncidentLogBuilder:
     def __init__(self, alert_group: "AlertGroup"):
         self.alert_group = alert_group
 
-    def get_log_records_list(
-        self, with_resolution_notes: bool = False
-    ) -> typing.List[typing.Union["AlertGroupLogRecord", "ResolutionNote", "UserNotificationPolicyLogRecord"]]:
+    def get_log_records_list(self, with_resolution_notes: bool = False) -> LogRecords:
```

Comment:
only references to `IncidentLogBuilder` in `grafana/oncall-private` are to this function (nothing has changed here)